### PR TITLE
Port VimL's method call syntax - vim-patch:8.1.{1638,1800,1803,1807,1809,1816,1820,1821,1828,1834,1835,1861,1863,1878,1879,1888,1909,1911,1912}

### PR DESCRIPTION
**Pull request by seandewar**
This is a WIP port of VimL's method call syntax: `base->method(...)`

Ref: #12525, #14611

I'm tempted to do this in multiple PRs to avoid a patch storm like #15211 to make reviewing this easier.
I was aiming for a relatively bug-free initial implementation in that PR which required _many_ patches.

However, method syntax seems to be quite bug-free from the get-go, with most patches only adding method call support to existing functions.

~~Note that I probably won't aim to support using `v:lua` as a method here.~~
Changed my mind 😄. `v:lua` may be used as a method like:
`base->v:lua.table.func(...)`

### CODE REVIEW

**Comment by seandewar** [comment]
@github-actions bad bot 👎

**Comment by dundargoc** [comment]
> @github-actions bad bot 👎 

What did it do?

**Comment by seandewar** [comment]
@dundargoc Nothing. I just didn't like its choice of labels is all 😄

**Comment by bfredl** [comment]
Maybe it should not add topic labels if someone already did that :)

**Comment by clason** [comment]
Yeah, it's (also) a bit eager with the `documentation` label (and often calls it a day even when there's actual code in the PR this documentation is about...)

File: src/nvim/eval.c [hunk 1]
```diff
@@ -8470,9 +8546,15 @@ handle_subscript(
       } else {
         s = (char_u *)"";
       }
-      ret = get_func_tv(s, lua ? slen : -1, rettv, (char_u **)arg,
-                        curwin->w_cursor.lnum, curwin->w_cursor.lnum,
-                        &len, evaluate, pt, selfdict);
+
+      funcexe_T funcexe;
+      memset(&funcexe, 0, sizeof(funcexe));
```
Comment (bfredl) [inline src/nvim/eval.c:8551]: Instead of memset, i would suggest defining `FUNCEXE_INIT`  in analogy with existing types.

**Comment by dundargoc** [comment]
> Maybe it should not add topic labels if someone already did that :)

Probably not worth it. The labeler is only triggered the first time a PR is opened. 

It takes around 10 seconds for the labels to be applied when I benchmarked it. Not only that, the author must be quick enough to label not only before the labeler is finished, but at the beginning of the labeler script where it's possible to check if there are any labels, so more like 5 seconds. I think it would only be relevant for label speedrunners.

The reason this particular PR didn't manage to label in time is because the GitHub servers tanked today for some reason. [Take a look at the recent workflow times](https://github.com/neovim/neovim/actions/workflows/labeler.yml). It took 25 min to label this PR. It's an anomaly that I don't think we should take into consideration for our workflow.

I propose what we do instead is refine and pinpoint when we want to apply labels. As I see it bot labeled correctly: the documentation files were modified so it added the `documentation` label.  The files under src/nvim/lua were also modified so the `lua` label was also applied. What makes these labels unsuited in this case as opposed to others when these files are modified? Is it because it's a vim-patch, and that vim-patch shouldn't activate the normal labels?

**Comment by bfredl** [comment]
@dundargoc every feature PR should add its own documentation. If "documentation" tag is to be added to every PR which is not a bugfix, the label is worthless.

**Comment by dundargoc** [comment]
> @dundargoc every feature PR should add its own documentation. If "documentation" tag is to be added to every PR which is not a bugfix, the label is worthless.

Agreed! That's why I think we should redefine when we want to use the documentation label. [I've created a PR for this](https://github.com/neovim/neovim/pull/15298), I propose we move the discussion there so @seandewar can work on his PR in peace :) .

File: src/nvim/testdir/test_functions.vim [hunk 2]
```diff
@@ -875,6 +875,34 @@ func Test_byte2line_line2byte()
   bw!
 endfunc
 
+func Test_byteidx()
```
Comment (seandewar) [inline src/nvim/testdir/test_functions.vim:878]: Ideally, how should this test be included?

It's a new-style version of a test in test69 that was already lua'd here: https://github.com/neovim/neovim/blob/35041432bff6812126e1cc04d30d35a5c9117537/test/functional/legacy/069_multibyte_formatting_spec.lua#L309

Difference though is that this test uses method call syntax, which is why I'd like to include it.

As the patch removed the old version of the test, should I remove the lua'd version?
Should I keep both versions of the test? Should I translate this to a lua test? :shrug:

Comment (bfredl) [reply src/nvim/testdir/test_functions.vim:878]: I'll keep both in this case. the lua test tests the actual functionality, and this one for the `->` shenanigans

**Comment by seandewar** [comment]
I think this is ready for review.

This PR doesn't include _all_ method-related patches.
The other patches (barring Vim9 stuff) add method support to existing built-in functions.
Those can be ported pretty easily in future PRs by editing `eval.lua` (I want to avoid making this PR too big).

Also, no Nvim-only functions have been made methods in this PR.

Btw, in the end I decided to add support for using `v:lua` as a method in this PR, e.g:

```vim
[[1], [2, 3], [4]]->v:lua.vim.tbl_flatten()->map({_, v -> v * 5})
" [5, 10, 15, 20]
```

**Comment by bfredl** [comment]
I haven't looked at everything, but the changes around `v:lua` LGTM.

**Review by justinmk — APPROVED** [review]
This is a thoughtful and careful port, thank you.

Just to be clear, we explicitly will not support Vimscript 2 (Vim9 vimscript). PRs adding such support will be rejected, we want to focus on Lua and editor features.

**Pull request merged**